### ISSUE TITLE
Why the batch type of the transformer is  token?

### ISSUE THREAD

**Issue by zhaoguangxiang** [OpenNMT/OpenNMT-py#906]
What is the benefit of setting the batch type to "token" ?
Thanks!

**Comment by sbmaruf** [comment]
From my experience,

Batch size considering token is a good consideration. The batch size makes training stable than the regular sentence wise batch type. The gradient update becomes more robust. Because in this way most of the batches become similar sized and also in seq2seq or transformer net we actually do the calculation on the basis of tokens. So the total number of token count should be more reasonable hyperparameter.

Imagine if your batch type is sentence. Your batch size is 32. for an arbitrary batch the average sentence size is 15 except that there is one sentence of 500. Now when you pad them the total batch size become 500*32=16000 tokens. Now if your batch size is `X`, different batch allocate a different amount of memory. This also increases the chance of OOM error.
